### PR TITLE
[base] Add exception module in base

### LINKED ISSUE #12: Find and debug unhandled exception in worker thread
When unhandled exception throws from worker thread, we cannot check the stack due to the design of std::thread. The reason for the unwound stack is that libstdc++'s function that calls the user's thread function contains a try/catch all statement.

The issue is related to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=55917. 

One solution is to use boost::thread to replace std::thread.

### PR DESCRIPTION
**Pull request by Yuzhen11**
Fix #12. Let Husky throw exception. To let the unhandled exception
easy to track, use boost::thread instead of std::thread.

### CODE REVIEW

File: core/job_runner.cpp [hunk 1]
```diff
@@ -18,6 +18,8 @@
 #include <thread>
 #include <vector>
 
+#include <boost/thread.hpp>
```
Comment (kygx-legend) [inline core/job_runner.cpp]: `#include "boost/thread.hpp"`

File: core/objlist.hpp [hunk 2]
```diff
@@ -19,6 +19,8 @@
 #include <unordered_map>
 #include <vector>
 
+#include "base/exception.hpp"
+
```
Comment (kygx-legend) [inline core/objlist.hpp]: No need to add empty line here.

**Comment by kygx-legend** [comment]
And please run `scripts/clang-format.sh` and commit again.

**Comment by Yuzhen11** [comment]
@kygx-legend  Please check again.

File: base/exception.hpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <exception>
+
```
Comment (kygx-legend) [inline base/exception.hpp]: lint.py: `base/exception.hpp:25:  Add #include <string> for string  [build/include_what_you_use] [4]`

**Pull request merged**